pkg/instance/instance_integration_test.go

```diff
@@ -61,68 +71,143 @@ func TestInstanceHandler(t *testing.T) {
-		instance.Routes(engine, func(ctx *gin.Context) {
-			ctx.Set("user", user)
-		}, instanceHandler)
+		instance.Routes(engine, authenticator, instanceHandler)
+
+		databaseHandler := database.NewHandler(databaseService, groupService, instanceService, stackService)
+		database.Routes(engine, authenticator, databaseHandler)
 	})
 
 	t.Run("Deploy", func(t *testing.T) {
 		var instance model.Instance
 		client.PostJSON(t, "/instances", strings.NewReader(`{
 			"name": "test-whoami",
-			"groupName": "test",
+			"groupName": "group-name",
 			"stackName": "whoami-go"
 		}`), &instance, inttest.WithAuthToken("sometoken"))
 
-		ctx, cancel := context.WithCancel(context.Background())
-		watch, err := k8sClient.Client.CoreV1().Pods(group.Name).Watch(ctx, metav1.ListOptions{
-			LabelSelector: "app.kubernetes.io/instance=" + instance.Name,
-		})
-		require.NoErrorf(t, err, "failed to find pod for instance %q", instance.Name)
-
-		timeout := 20 * time.Second
-		tm := time.NewTimer(timeout)
-		defer tm.Stop()
-		for {
-			select {
-			case <-tm.C:
-				assert.Fail(t, "timed out waiting on pod")
+		assertPodIsReady(t, k8sClient, group.Name, instance.Name)
+	})
+
+	var databaseID string
+	{
+		t.Log("Upload")
+		var b bytes.Buffer
+		w := multipart.NewWriter(&b)
+		err := w.WriteField("group", "group-name")
+		require.NoError(t, err, "failed to write form field")
+		f, err := w.CreateFormFile("database", "mydb")
+		require.NoError(t, err, "failed to create form file")
+		// TODO: Why does this work?
+		_, err = io.WriteString(f, "file contents")
+		require.NoError(t, err, "failed to write file")
+		_ = w.Close()
+
+		body := client.Post(t, "/databases", &b, inttest.WithHeader("Content-Type", w.FormDataContentType()))
+
+		var actualDB model.Database
+		err = json.Unmarshal(body, &actualDB)
+		require.NoError(t, err, "POST /databases: failed to unmarshal HTTP response body")
+		require.Equal(t, "mydb", actualDB.Name)
+		require.Equal(t, "group-name", actualDB.GroupName)
+
+		actualContent := s3.GetObject(t, s3Bucket, "group-name/mydb")
+		require.Equalf(t, "file contents", string(actualContent), "DB in S3 should have expected content")
+
+		databaseID = strconv.FormatUint(uint64(actualDB.ID), 10)
+	}
+
+	t.Run("DeployDHIS2", func(t *testing.T) {
+		t.Setenv("HOSTNAME", "test-dhis2.group-name.svc")
+
+		var instance model.Instance
+		client.PostJSON(t, "/instances", strings.NewReader(`{
+			"name": "test-dhis2",
+			"groupName": "group-name",
+			"stackName": "dhis2",
+			"parameters": [
+                {
+    		        "name": "DATABASE_ID",
+			        "value": "`+databaseID+`"
+			    }
+			]
+		}`), &instance, inttest.WithAuthToken("sometoken"))
+
+		assertPodIsReady(t, k8sClient, group.Name, instance.Name+"-database")
+		assertPodIsReady(t, k8sClient, group.Name, instance.Name)
+	})
+}
+
+func assertPodIsReady(t *testing.T, k8sClient *inttest.K8sClient, group string, instance string) {
```

Comment:
If we are repeatedly using our inttest client to do the same thing I would make this a method on the inttest client

```go
k8sClient := inttest.SetupK8s(t)
...
k8sClient.WaitUntilPodIsReady(t, group, instance)
```

If a test is doing something very specific they can then opt-out of our inttest client and use the underlying one like k8sClient.Client (or S3, ...).

Reply:
Currently we're only using it in one test. It'll probably be used in some of the database tests as well but since that's not the case now I just placed in the instance test.

Despite the above I've been thinking if we should move the save, and save as, functionality into the instance handler so we have endpoints such as /instance/:id/save rather than /save/:instanceId on the database handler but that comes with some issues which I wanted to discuss with you.

Reply:
> Currently we're only using it in one test.
You are using the function 3 times which to me is good enough to move this to the inttest client 😄

Reply:
Sure! Should we make it a member of the K8sClient?